### PR TITLE
fix skipped patch if source package no git index

### PR DESCRIPTION
**Pull request by ivanallen**
* Before adding new features and new modules, please go to issues to submit the relevant feature description first.
* Write good commit messages and use the same coding conventions as the rest of the project.
* Please commit code to dev branch and we will merge into master branch in feature
* Ensure your edited codes with four spaces instead of TAB.

------

* 增加新特性和新模块之前，请先到issues提交相关特性说明，经过讨论评估确认后，再进行相应的代码提交，避免做无用工作。
* 编写友好可读的提交信息，并使用与工程代码相同的代码规范，代码请用4个空格字符代替tab缩进。
* 请提交代码到dev分支，如果通过，我们会在特定时间合并到master分支上。
* 为了规范化提交日志的格式，commit消息，不要用中文，请用英文描述。

#4356

### CODE REVIEW

File: xmake/modules/private/action/require/impl/actions/patch_sources.lua [hunk 1]
```diff
@@ -98,7 +99,9 @@ function _patch(package, patch_url, patch_hash)
     end
 
     -- apply the patch file
-    git.apply(patch_file)
+    local patch = assert(find_tool("patch"), "patch not found!")
```
Comment (SirLynix) [inline xmake/modules/private/action/require/impl/actions/patch_sources.lua:102]: this will break on platforms where patch isn't builtin, like windows

**Comment by SirLynix** [comment]
I don't understand why this is needed since `git apply` works on non-git folder. It does work when downloading a .tar.gz and then applying a patch.

example with the [chipmunk](https://github.com/xmake-io/xmake-repo/blob/dev/packages/c/chipmunk2d/xmake.lua) package:

```
note: install or modify (m) these packages (pass -y to skip confirm)?
in xmake-repo:
  -> chipmunk2d 7.0.3 [vs_runtime:"MDd"]
please input: y (y/n/m)
y
pinging the host(github.com) ... 22 ms
C:\Program Files\xmake\winenv\bin\7z x -y Chipmunk-7.0.3.tar.gz -oC:\Users\lynix\AppData\Local\Temp\.xmake\231103\_0ACAF9EAD117437087443578F521AA90.tar

7-Zip 19.00 (x64) : Copyright (c) 1999-2018 Igor Pavlov : 2019-02-21

Scanning the drive for archives:
1 file, 514376 bytes (503 KiB)

Extracting archive: Chipmunk-7.0.3.tar.gz
--
Path = Chipmunk-7.0.3.tar.gz
Type = gzip
Headers Size = 10

Everything is Ok

Size:       2426880
Compressed: 514376
C:\Program Files\xmake\winenv\bin\7z x -y C:\Users\lynix\AppData\Local\Temp\.xmake\231103\_0ACAF9EAD117437087443578F521AA90.tar\Chipmunk-7.0.3.tar -osource.tmp

7-Zip 19.00 (x64) : Copyright (c) 1999-2018 Igor Pavlov : 2019-02-21

Scanning the drive for archives:
1 file, 2426880 bytes (2370 KiB)

Extracting archive: C:\Users\lynix\AppData\Local\Temp\.xmake\231103\_0ACAF9EAD117437087443578F521AA90.tar\Chipmunk-7.0.3.tar
--
Path = C:\Users\lynix\AppData\Local\Temp\.xmake\231103\_0ACAF9EAD117437087443578F521AA90.tar\Chipmunk-7.0.3.tar
Type = tar
Physical Size = 2426880
Headers Size = 121344
Code Page = UTF-8

Everything is Ok

Folders: 26
Files: 192
Size:       2257693
Compressed: 2426880
patching C:\Users\lynix\AppData\Local\.xmake\repositories\xmake-repo\packages\c\chipmunk2d\patches\7.0.3\android.patch to chipmunk2d-7.0.3 ..
git apply --reject --ignore-whitespace C:\Users\lynix\AppData\Local\Temp\.xmake\231103\patches\chipmunk2d\7.0.3\android.patch
Checking patch include/chipmunk/chipmunk_types.h...
Checking patch src/cpHastySpace.c...
Applied patch include/chipmunk/chipmunk_types.h cleanly.
Applied patch src/cpHastySpace.c cleanly.
```

image

this folder has no .git hidden folder and git apply works well.

**Comment by ivanallen** [comment]
> I don't understand why this is needed since `git apply` works on non-git folder. It does work when downloading a .tar.gz and then applying a patch.
> 
> example with the [chipmunk](https://github.com/xmake-io/xmake-repo/blob/dev/packages/c/chipmunk2d/xmake.lua) package:
> 
> ```
> note: install or modify (m) these packages (pass -y to skip confirm)?
> in xmake-repo:
>   -> chipmunk2d 7.0.3 [vs_runtime:"MDd"]
> please input: y (y/n/m)
> y
> pinging the host(github.com) ... 22 ms
> C:\Program Files\xmake\winenv\bin\7z x -y Chipmunk-7.0.3.tar.gz -oC:\Users\lynix\AppData\Local\Temp\.xmake\231103\_0ACAF9EAD117437087443578F521AA90.tar
> 
> 7-Zip 19.00 (x64) : Copyright (c) 1999-2018 Igor Pavlov : 2019-02-21
> 
> Scanning the drive for archives:
> 1 file, 514376 bytes (503 KiB)
> 
> Extracting archive: Chipmunk-7.0.3.tar.gz
> --
> Path = Chipmunk-7.0.3.tar.gz
> Type = gzip
> Headers Size = 10
> 
> Everything is Ok
> 
> Size:       2426880
> Compressed: 514376
> C:\Program Files\xmake\winenv\bin\7z x -y C:\Users\lynix\AppData\Local\Temp\.xmake\231103\_0ACAF9EAD117437087443578F521AA90.tar\Chipmunk-7.0.3.tar -osource.tmp
> 
> 7-Zip 19.00 (x64) : Copyright (c) 1999-2018 Igor Pavlov : 2019-02-21
> 
> Scanning the drive for archives:
> 1 file, 2426880 bytes (2370 KiB)
> 
> Extracting archive: C:\Users\lynix\AppData\Local\Temp\.xmake\231103\_0ACAF9EAD117437087443578F521AA90.tar\Chipmunk-7.0.3.tar
> --
> Path = C:\Users\lynix\AppData\Local\Temp\.xmake\231103\_0ACAF9EAD117437087443578F521AA90.tar\Chipmunk-7.0.3.tar
> Type = tar
> Physical Size = 2426880
> Headers Size = 121344
> Code Page = UTF-8
> 
> Everything is Ok
> 
> Folders: 26
> Files: 192
> Size:       2257693
> Compressed: 2426880
> patching C:\Users\lynix\AppData\Local\.xmake\repositories\xmake-repo\packages\c\chipmunk2d\patches\7.0.3\android.patch to chipmunk2d-7.0.3 ..
> git apply --reject --ignore-whitespace C:\Users\lynix\AppData\Local\Temp\.xmake\231103\patches\chipmunk2d\7.0.3\android.patch
> Checking patch include/chipmunk/chipmunk_types.h...
> Checking patch src/cpHastySpace.c...
> Applied patch include/chipmunk/chipmunk_types.h cleanly.
> Applied patch src/cpHastySpace.c cleanly.
> ```
> 
> image
> 
> this folder has no .git hidden folder and git apply works well.

I don't know but it's not working on linux...

**Comment by SirLynix** [comment]
> it's not working on linux...

it does work on Linux:
```
lynix@Malcolm:~/ThisSpaceOfMine$ xmake require -f --shallow -vD chipmunk2d
/usr/bin/git rev-parse HEAD
note: install or modify (m) these packages (pass -y to skip confirm)?
in xmake-repo:
  -> chipmunk2d 7.0.3
please input: y (y/n/m)
y
pinging for the host(github.com) ... 7 ms
checking for curl ... /usr/bin/curl
/usr/bin/curl -SL -A "Xmake/2.8.1+HEAD.12fe4e0 (Linux;5.15.0-88-generic) curl/7.81.0" https://github.com/slembcke/Chipmunk2D/archive/Chipmunk-7.0.3.tar.gz -o Chipmunk-7.0.3.tar.gz
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
  0     0    0     0    0     0      0      0 --:--:-- --:--:-- --:--:--     0
100  502k    0  502k    0     0   766k      0 --:--:-- --:--:-- --:--:--  766k
/usr/bin/tar -xf Chipmunk-7.0.3.tar.gz -C source.tmp
  => download https://github.com/slembcke/Chipmunk2D/archive/Chipmunk-7.0.3.tar.gz .. ok
patching /home/lynix/.xmake/repositories/xmake-repo/packages/c/chipmunk2d/patches/7.0.3/android.patch to chipmunk2d-7.0.3 ..
/usr/bin/git apply --reject --ignore-whitespace /tmp/.xmake1001/231103/patches/chipmunk2d/7.0.3/android.patch
Checking patch include/chipmunk/chipmunk_types.h...
Checking patch src/cpHastySpace.c...
Applied patch include/chipmunk/chipmunk_types.h cleanly.
Applied patch src/cpHastySpace.c cleanly.
```

**Comment by ivanallen** [comment]
> > it's not working on linux...
> 
> it does work on Linux as well:
> 
> ```
> lynix@Malcolm:~/ThisSpaceOfMine$ xmake require -f --shallow -vD chipmunk2d
> /usr/bin/git rev-parse HEAD
> note: install or modify (m) these packages (pass -y to skip confirm)?
> in xmake-repo:
>   -> chipmunk2d 7.0.3
> please input: y (y/n/m)
> y
> pinging for the host(github.com) ... 7 ms
> checking for curl ... /usr/bin/curl
> /usr/bin/curl -SL -A "Xmake/2.8.1+HEAD.12fe4e0 (Linux;5.15.0-88-generic) curl/7.81.0" https://github.com/slembcke/Chipmunk2D/archive/Chipmunk-7.0.3.tar.gz -o Chipmunk-7.0.3.tar.gz
>   % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
>                                  Dload  Upload   Total   Spent    Left  Speed
>   0     0    0     0    0     0      0      0 --:--:-- --:--:-- --:--:--     0
> 100  502k    0  502k    0     0   766k      0 --:--:-- --:--:-- --:--:--  766k
> /usr/bin/tar -xf Chipmunk-7.0.3.tar.gz -C source.tmp
>   => download https://github.com/slembcke/Chipmunk2D/archive/Chipmunk-7.0.3.tar.gz .. ok
> patching /home/lynix/.xmake/repositories/xmake-repo/packages/c/chipmunk2d/patches/7.0.3/android.patch to chipmunk2d-7.0.3 ..
> /usr/bin/git apply --reject --ignore-whitespace /tmp/.xmake1001/231103/patches/chipmunk2d/7.0.3/android.patch
> Checking patch include/chipmunk/chipmunk_types.h...
> Checking patch src/cpHastySpace.c...
> Applied patch include/chipmunk/chipmunk_types.h cleanly.
> Applied patch src/cpHastySpace.c cleanly.
> ```

I already know. Because I set the .xmake global dir in my project
```
- myprojject_dir
   |- .git  # this is myproject git repo
   |- xmake_global_dir
        |- .xmake
             |- .cache
                ...
                  |- grpc_source  # git apply not working, because git found the repo is myproject
        
```

**Comment by SirLynix** [comment]
maybe just passing --git-dir=. would fix this

**Comment by ivanallen** [comment]
> maybe just passing --git-dir=. to git apply would fix this

image


It's not working
image

**Comment by SirLynix** [comment]
maybe passing the current absolute folder as --directory would work, but it's mostly a hack

**Comment by ivanallen** [comment]
> maybe passing the current absolute folder as --directory would work, but it's mostly a hack

I already tried it and it didn't work

**Comment by waruqi** [comment]
It works for me. 

```console
$ git apply --reject --ignore-whitespace /tmp/patch
Skipped patch 'update1.c'.
$ git --git-dir=. apply  --reject --ignore-whitespace /tmp/patch
Checking patch update1.c...
Applied patch update1.c cleanly.
```

**Comment by ivanallen** [comment]
@SirLynix I mistakenly thought git-dir meant the parameter of apply. It works now.

**Review by SirLynix — APPROVED** [review]

File: xmake/modules/devel/git/apply.lua [hunk 2]
```diff
@@ -38,6 +38,6 @@ import("lib.detect.find_tool")
 function main(patchfile, opt)
     opt = opt or {}
     local git = assert(find_tool("git"), "git not found!")
-    local argv = {"apply", "--reject", "--ignore-whitespace", patchfile}
+    local argv = {"--git-dir=.", "apply", "--reject", "--ignore-whitespace", patchfile}
```
Comment (waruqi) [inline xmake/modules/devel/git/apply.lua:41]: please pass gitdir argument.

```lua
if opt.gitdir then
    table.insert(argv, 1, opt.gitdir)
end
```

Comment (waruqi) [reply xmake/modules/devel/git/apply.lua:41]: and pass `git.apply(file, {gitdir = os.curdir()})`

**Comment by waruqi** [comment]
@ivanallen It will break something. https://github.com/xmake-io/xmake-repo/pull/2781

**Comment by waruqi** [comment]
I revert this first.

**Comment by ivanallen** [comment]
> I revert this first.

sad

**Pull request merged**